src/core/model.js

```diff
@@ -65,30 +77,25 @@ export default class Model extends Object3D {
     // Other opts
     timerQueryEnabled = false
   } = {}) {
-    const {gl} = this;
-
-    if (!vs) {
-      vs = modules ? MODULAR_SHADERS.vs : MONOLITHIC_SHADERS.vs;
-    }
-    if (!fs) {
-      fs = modules ? MODULAR_SHADERS.fs : MONOLITHIC_SHADERS.fs;
-    }
-
-    // Assign default uniforms if any default shaders are being used
-    if (vs === MONOLITHIC_SHADERS.vs || fs === MONOLITHIC_SHADERS.fs) {
-      defaultUniforms = defaultUniforms || MONOLITHIC_SHADERS.defaultUniforms;
-    }
+    this._initializeProgram({
+      vs,
+      fs,
+      modules,
+      defines,
+      moduleSettings,
+      defaultUniforms,
+      program
+    });
 
-    // Call assembleShaders if `modules` were supplied
-    let getUniforms;
-    if (modules && typeof vs === 'string' && typeof fs === 'string') {
-      ({vs, fs, getUniforms} = assembleShaders(gl, {vs, fs, modules}));
-    }
-    this.getModuleUniforms = getUniforms || (x => {});
+    this.uniforms = {};
 
-    // set a custom program per o3d
-    this.program = program || new Program(gl, {vs, fs});
-    assert(this.program instanceof Program, 'Model needs a program');
+    // Make sure we have some reasonable default uniforms in place
+    uniforms = Object.assign({}, this.program.defaultUniforms, uniforms);
+    this.setUniforms(uniforms);
```

Comment:
Agreed - we have too many `setUniforms` on Program, Model, Layer. We should take a good look at them and their semantics and it could be that we can find a better naming.